core/src/main/java/com/nisovin/magicspells/spells/passive/InputListener.java

```diff
@@ -0,0 +1,178 @@
+
+	private InputPredicate oldInputPredicate;
+	private InputPredicate newInputPredicate;
+
+	@Override
+	public void initialize(@NotNull String var) {
+		MagicSpells.error("PassiveSpell '" + passiveSpell.getInternalName() + "' attempted to create a 'input' trigger using the string format, which it does not support.");
+	}
+
+	@Override
+	public boolean initialize(@NotNull ConfigurationSection config) {
+		List<String> onPressStrings = config.getStringList("on-press");
+		for (String typeString : onPressStrings) {
+			InputType type;
+			try {
+				type = InputType.valueOf(typeString.toUpperCase());
+			} catch (IllegalArgumentException e) {
+				MagicSpells.error("Invalid input type '" + typeString + "' specified in 'on-press' in 'input' trigger on passive spell '" + passiveSpell.getInternalName() + "'.");
+				return false;
+			}
+
+			onPress.add(type);
+		}
+
+		List<String> onReleaseStrings = config.getStringList("on-release");
+		for (String typeString : onReleaseStrings) {
+			InputType type;
+			try {
+				type = InputType.valueOf(typeString.toUpperCase());
+			} catch (IllegalArgumentException e) {
+				MagicSpells.error("Invalid input type '" + typeString + "' specified in 'on-release' in 'input' trigger on passive spell '" + passiveSpell.getInternalName() + "'.");
+				return false;
+			}
+
+			onRelease.add(type);
+		}
+
+		String oldInputString = config.getString("old-input");
+		if (oldInputString != null) {
+			oldInputPredicate = InputPredicate.fromString(oldInputString);
+
+			if (oldInputPredicate == null) {
+				MagicSpells.error("Invalid value '" + oldInputString + "' specified for 'old-input' in 'input' trigger on passive spell '" + passiveSpell.getInternalName() + "'.");
+				return false;
+			}
+		}
+
+		String newInputString = config.getString("new-input");
+		if (newInputString != null) {
+			newInputPredicate = InputPredicate.fromString(newInputString);
+
+			if (newInputPredicate == null) {
+				MagicSpells.error("Invalid value '" + newInputString + "' specified for 'new-input' in 'input' trigger on passive spell '" + passiveSpell.getInternalName() + "'.");
+				return false;
+			}
+		}
+
+		return true;
+	}
+
+	@OverridePriority
+	@EventHandler
+	public void onInput(PlayerInputEvent event) {
+		Player caster = event.getPlayer();
+		if (!canTrigger(caster)) return;
+
+		Input oldInput = caster.getCurrentInput();
+		if (oldInputPredicate != null && !oldInputPredicate.test(oldInput)) return;
+
+		Input newInput = event.getInput();
+		if (newInputPredicate != null && !newInputPredicate.test(newInput)) return;
+
+		trigger_check:
+		if (!onPress.isEmpty() || !onRelease.isEmpty()) {
+			for (InputType type : onPress)
+				if (!type.isPressed(oldInput) && type.isPressed(newInput))
+					break trigger_check;
+
+			for (InputType type : onRelease)
+				if (type.isPressed(oldInput) && !type.isPressed(newInput))
+					break trigger_check;
+
+			return;
+		}
+
+		passiveSpell.activate(caster);
+	}
+
+	private enum InputType {
```

Comment:
You could reduce this quite a bit with a constructor accepting a predicate lambda